### ISSUE TITLE
inTrees for ranger (fast implementation of RandomForest)

### ISSUE THREAD

**Issue by shivonkar** [softwaredeng/inTrees#3]
The "ranger" is the fast implementation of RandomForest. I have been using ranger for last 3 years and found it better in terms of speed and accuracy then core "randomForest" package. Please see if you can implement for "ranger" too

**Comment by mnwright** [comment]
We now have the function `treeInfo()` in ranger, see imbs-hl/ranger#245. Does this help? Do you need anything else for the interface?

Example: 
````R
devtools::install_github("imbs-hl/ranger") # Not on CRAN yet
library(ranger)
rf <- ranger(Species ~ ., data = iris)
treeInfo(rf, tree = 1)
````

**Comment by shivonkar** [comment]
Thank you for providing. How can I get the tree in follwoing scenario. 

Let us say, you have new test data

   Sepal.Length Sepal.Width Petal.Length Petal.Width
           5.2         3.6               1.5                 0.3

How to get the tree (may be ensembled tree), that was used to predict the outcome (Species)

**Comment by shivonkar** [comment]
Hi Marvin,
Thank you for following up

It did not solve the issue. I have put the comment at link
https://github.com/softwaredeng/inTrees/issues/3

Thank you
Shiv, PhD

**Comment by mnwright** [comment]
The function `treeInfo()` in ranger just helps to get similar info as with `getTree()` in randomForest. The rest is up to the developers of inTrees.

**Comment by softwaredeng** [comment]
thanks @mnwright! I just saw this update. Is it on CRAN now (if not any plan for that)?  cc @shivonkar

**Comment by mnwright** [comment]
Yes, it's on CRAN since version 0.9.0.

**Comment by softwaredeng** [comment]
@mnwright I have almost finished the integration and there is just one thing to confirm with you. For a categorical data set, there are around 300 out of 4000 paths/conditions (path is from a root node to a leaf node, e.g., X1 = L1 & X2 = L2 ) are not valid, that is, no data point satisfies any of these 300 rules. the original RF version doesn't have this issue. Do you think it is how Ranger works, or something may be wrong?

**Comment by mnwright** [comment]
This means some terminal nodes are empty? Could you share the example?

**Comment by softwaredeng** [comment]
@mnwright I believe the rules are wrongly extracted. I did another check, and suspecting splitval may be the issue. I have created an R space with the result: https://github.com/softwaredeng/inTrees/blob/master/intrees_ranger_debug.RData 
and source code with comment is here: https://github.com/softwaredeng/inTrees/blob/master/debug_intrees_ranger.R

let me know more inf is needed.

**Comment by mnwright** [comment]
It looks like there are observations in that path. If ranger is called with `respect.unordered.factors = "ignore"` (the default), the factors are treated as numeric in the order of the factor levels. That is, a split value of 2.5 in X3 in your example will be that " all-paid-duly"  and " bank-paid-duly" go to the left, the others to the right. 
If I'm not mistaken, the rule should be 
`which(X[,1] %in% c("0DM", 'less-200DM') & X[,3] %in% c(' bank-paid-duly', " all-paid-duly") & X[,6] %in% c(" less1000DM", ' less100DM') & X[,12] %in% c(" building-society", ' car') & X[,13]<=51)
`
which isn't empty.

**Comment by softwaredeng** [comment]
@mnwright Got it. if the factors un-ordered, then we would have an issue? 

How do you think using what randomForests package does? that is applicable to both ordered or unordered categorical variables. 

For categorical predictors, the splitting point is represented by an integer, whose binary expansion
gives the identities of the categories that goes to left or right. For example, if a predictor has
four categories, and the split point is 13. The binary expansion of 13 is (1, 0, 1, 1) (because
13 = 1 ∗ 2^0+ 0 ∗ 2^1 + 1 ∗ 2^2 + 1 ∗ 2^3), so cases with categories 1, 3, or 4 in this predictor get sent
to the left, and the rest to the right.

**Comment by mnwright** [comment]
Yes, unordered factors are handled differently. Internally, we are doing the same thing as in randomForest, but `treeInfo()` resolves this. From `?treeInfo`:

> In the "partition" mode, the splitval values for unordered factor are comma separated lists of values, representing the factor levels (in the original order) going to the left.

Try for example
````R

library(ranger)
n <- 50
dt <- data.frame(x = sample(LETTERS[1:10], n, replace = TRUE), 
                 y = rbinom(n, 1, 0.5), 
                 stringsAsFactors = FALSE)

rf <- ranger(y ~ x, dt, respect.unordered.factors = "partition")
treeInfo(rf, 1)
````

**Comment by softwaredeng** [comment]
@mnwright got it. so there are two ways handling categorical variables: 
1. ordered, which is treated as 1,2,3,4...2. 
2. unordered. which uses a string separated by ","

Now from the ranger object, how do we know which value of respect.unordered.factors was used? If there is no way,  we can ask users to specify the value used in order to extract value.

**Comment by mnwright** [comment]
Exactly. That can be checked with `rf$forest$is.ordered`, which is `FALSE` for any variable splitted with the partition approach. Example: 

````R
library(ranger)
rf <- ranger(Sepal.Length ~ ., iris, respect.unordered.factors = "order")
rf$forest$is.ordered # All TRUE

rf <- ranger(Sepal.Length ~ ., iris, respect.unordered.factors = "partition")
rf$forest$is.ordered # One FALSE
```

**Comment by softwaredeng** [comment]
This question doesn't impact how we extract rules, just curious. How does ranger decides which one is ordered? when I used the option of respect.unordered.factors = "partition", some categorical variables still have is.ordered = TRUE.

**Comment by mnwright** [comment]
Could you give an example?

**Comment by aruaud** [comment]
Hi! Great package and thanks for adding ranger to the implementation!
In `Ranger2List()` and `formatRanger()`, the `ranger::treeInfo()` returns the `tree$splitvarID` indexed from 0 instead of 1, leading `treeVisit()` to then fail. 
Could you add +1 in `formatRanger()` as for the nodes? That should solve issues #3 and #15.

``` 
Ranger2List <- function(rf_ranger)
{
  formatRanger <- function(tree){
   rownames(tree) <- 1:nrow(tree)
   tree$`status` <- ifelse(tree$`terminal`==TRUE,-1,1)
   tree$`left daughter` <- tree$`leftChild` + 1
   tree$`right daughter` <- tree$`rightChild` + 1
   tree$`split var` <- tree$`splitvarID` + 1 

   etc
  }
}
```

**Comment by softwaredeng** [comment]
Thank you @aruaud! This seems to work! Will combine this and some minor changes to release a new version.
